### PR TITLE
Crafting round-trip test for Cff importer/exporter

### PR DESCRIPTION
**Pull request by Raahitya-14**
This PR adds a round-trip test for Cff importer and exporter as requested in the comments of #10661. For the round-trip test to work, we need to address the difference in implementations between the importer and the exporter. In cff, there is one type of field, which is called 'preferred-citation' that can be used to give different types than 'sofware' to your entry, the exporter always uses it to pass the relevant type (i.e. 'article', 'inproceedings', 'book', etc) but the importer does not parse this field. Therefore, to address this issue and for the round-trip test to work, modification of the importer to parse the 'preferred-citation' field was required. There also is a problem with the author's name which needs to be changed to accept two arguments. 

 ### Mandatory checks
 * [ ]  Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
 * [x]  Tests created for changes (if applicable)
 * [ ]  Manually tested changed features in running JabRef (always required)
 * [ ]  Screenshots added in PR description (for UI changes)
 * [ ]  [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
 * [ ]  [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at https://github.com/JabRef/user-documentation/issues or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/test/java/org/jabref/logic/importer/CffImporterTest.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+            assertEquals("My awesome research software", entry.getField(StandardField.TITLE).orElse(null));
+            assertEquals("Mona Lisa and Hew Bot", entry.getField(StandardField.AUTHOR).orElse(null));
+            assertEquals("10.0000/00000", entry.getField(StandardField.DOI).orElse(null));
+            assertEquals("Journal Title", entry.getField(StandardField.JOURNAL).orElse(null));
+            assertEquals("1", entry.getField(StandardField.VOLUME).orElse(null));
+            assertEquals("1", entry.getField(StandardField.ISSUE).orElse(null));
+            assertEquals("1-10", entry.getField(StandardField.PAGES).orElse(null));
+            assertEquals("2021", entry.getField(StandardField.YEAR).orElse(null));
+            System.out.println("Test Successful!");
```
Comment (koppor) [inline src/test/java/org/jabref/logic/importer/CffImporterTest.java:69]: Please rewrite using the idea of org.jabref.logic.database.DatabaseMergerTest#mergeAddsNonDuplicateEntries: List.of, BibEntry.withfield...

Please no Println with "test successful". IntelliJ will have green or red indicators. See the last part (marked as "optional") in our guide: https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html

**Review by koppor — COMMENTED** [review]
The diff is too large for me to inspect wihtout tool support. 

Until I manage to get a AST-based diff tool running, a small comment to keep you going.

File: src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java [hunk 2]
```diff
@@ -92,73 +103,99 @@ public CffIdentifier() {
...
+        MainCffContentMethod(citation, entryMap, entry);
+        entryMap.forEach(entry::setField);
 
         List<BibEntry> entriesList = new ArrayList<>();
         entriesList.add(entry);
 
         return new ParserResult(entriesList);
     }
 
+    private void PreferredCitationMethod(JsonNode preferredCitation, Map<Field, String> entryMap, BibEntry entry) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:127]: method start with a lower case letter (lower camel case)

File: src/test/java/org/jabref/logic/importer/CffImporterTest.java (hunk 1 above)
Comment (Raahitya-14) [reply src/test/java/org/jabref/logic/importer/CffImporterTest.java:69]: Sure , I'll make the necessary changes.

File: src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java (hunk 2 above)
Comment (Raahitya-14) [reply src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:127]: Sure I'll change the method name. Thank you!

**Comment by koppor** [comment]
@Raahitya-14 Please do not reference issues in commits. The pull request should reference an issue. Reasoning: There will be more than 1 commit referencing an issue. Thus, the issue will be swamped with references to commits, which are becoming obsolete while the implementation evolves.

Round trip was asked for at https://github.com/JabRef/jabref/pull/10917#issuecomment-1964906618. Thus, the, referenced issue https://github.com/JabRef/jabref/issues/10661 was not right.

- Import was implemented at https://github.com/JabRef/jabref/pull/7946 (Issue https://github.com/JabRef/jabref/issues/7945)
- Export was implemented at https://github.com/JabRef/jabref/pull/10917 (Issue https://github.com/JabRef/jabref/issues/10661)

File: src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java [hunk 3]
```diff
@@ -92,73 +103,100 @@ public CffIdentifier() {
...
-            }
+        StandardEntryType entryType = StandardEntryType.Misc;
+
+        if (citation.type != null) {
+            entryType = mapType(citation.type);
         }
+        BibEntry entry = new BibEntry(entryType);
+        HashMap<Field, String> entryMap = new HashMap<>();
 
-        // Select SWHID to keep
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:133]: I don't see this in the new code.

Isn't there a test case for that? Yes -- org.jabref.logic.importer.fileformat.CffImporterTest#importEntriesSwhIdSelect1

File: src/test/java/org/jabref/logic/importer/CffImporterTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import org.jabref.logic.importer.ParserResult;
+import org.jabref.logic.importer.fileformat.CffImporter;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.field.StandardField;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class CffImporterTest {
```
Comment (koppor) [inline src/test/java/org/jabref/logic/importer/CffImporterTest.java:14]: This class resides in the wrong package. Didn't you use Ctrl+Shift+T as outlined in our guidelines?

You will note that there is already a `org/jabref/logic/importer/fileformat/CffImporterTest.java`. Please add your additional tests there.

File: src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java [hunk 5]
```diff
@@ -92,73 +103,100 @@ public CffIdentifier() {
...
+        if (!entryMap.containsKey(StandardField.TITLE) && citation.values.containsKey("title")) {
+            entryMap.put(StandardField.TITLE, citation.values.get("title"));
+        }
+        if (!entryMap.containsKey(StandardField.AUTHOR) && citation.authors != null && !citation.authors.isEmpty()) {
+            List<String> authorsList = new ArrayList<>();
+            for (CffAuthor author : citation.authors) {
+                String givenName = author.values.getOrDefault("given-names", "");
+                String familyName = author.values.getOrDefault("family-names", "");
+                authorsList.add((givenName + " " + familyName).trim());
+            }
```
Comment (koppor) [inline src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:180]: Seems to be code duplication compared to above. But also slightly different. Maybe a code comment should be added why the code at `preferredCitationMethode` is different from here.

**Review by koppor — CHANGES_REQUESTED** [review]
Commenting PRs with guidance without fully thinking through is a though work - when the own time limits are respected.

I think, the direction is OKish, but you should try to keep existing functionality, which is perfectly backed by tests. Therefore, you see failing unit tests at `org.jabref.logic.importer.fileformat.CffImporterTest`. Please update the code to cover the existing use case

---

Work on importing https://github.com/JabRef/jabref/blob/main/CITATION.cff

---

In case there is a preferred citation, the other citation should be imported to **another** BibEntry. (I am not sure if the infrastructuire of JabRef supports that, but it should be able to generated muliple entries based on a single file)

**Comment by koppor** [comment]
@Raahitya-14 Please keep watching the red crosses of GitHub and try to fix them:

image

**Pull request closed without merging**